Unifying ranges

OlivierBBB opened the pull request.

OlivierBBB committed:
first commit

OlivierBBB committed:
more wip

OlivierBBB committed:
refactoring done ?

OlivierBBB committed:
spotless

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/Util.java:
```diff
@@ -168,25 +168,45 @@ public static short getTxTypeAsInt(TransactionType txType) {
-
-    if (dataSize >= positionStart) {
-      if (dataSize >= (positionStart + size)) {
-        output = data.slice(positionStart, size);
```
The condition `dataSize >= positionStart` is problematic as `Bytes slice(int i, int length);` throws an exception if `i >= length`. Also some simple reformatting.

OlivierBBB committed:
ras: formatting + documentation + static imports

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/types/MemoryRange.java:
The main change in this PR. The goal is to replace all ad-hoc ranges with a `MemoryRange`.

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/types/Range.java:
ras: renaming from `MemorySpan -> Range`. Two reasons:
- a **range** is a minimum value and a maximum value while a **span** is the difference between the two
- I want to reserve the word `MemoryRange` to something that actually points into _someone's_ memory

OlivierBBB replied:
<img width="862" alt="image" src="https://github.com/user-attachments/assets/8715cd8a-b2cb-4138-a9de-1aa2017edff0">

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/types/MemoryRange.java:
```diff
@@ -12,53 +12,103 @@
+    return range.isEmpty()
+        ? Bytes.EMPTY
+        : slice(rawData, safeLongToInt(range.offset()), safeLongToInt(range.size()));
   }
```
This method e.g. extracts the relevant data from the `rawData` e.g.
- `callData` from the `callerMemorySnapshot`
- `returnData` from the `callerMemorySnapshot`

OlivierBBB committed:
ras

OlivierBBB committed:
ras

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/Util.java:
```diff
@@ -168,25 +168,45 @@ public static short getTxTypeAsInt(TransactionType txType) {
+   * @return
+   * @throws ArithmeticException
+   */
+  public static int safeLongToInt(long value) throws ArithmeticException {
```
Move this fct in conversions.java

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/Util.java:
```diff
@@ -168,25 +168,45 @@ public static short getTxTypeAsInt(TransactionType txType) {
+   * @param offset
    */
-  public static Bytes slice(Bytes data, int positionStart, int size) {
+  public static Bytes slice(Bytes data, int offset, int size) {
```
I was planning to do it in my PR. I'll have to resolve conflicts now 😅

OlivierBBB committed:
more wip

OlivierBBB committed:
spotless

OlivierBBB committed:
wip

OlivierBBB committed:
spotless

OlivierBBB committed:
more stuff

OlivierBBB committed:
ras

OlivierBBB committed:
ras

OlivierBBB committed:
fix: some EXTCODECOPY fixes

letypequividelespoubelles committed:
fix modexp raw lead

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
constraints update

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
constraints

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
fix(modexp): rawLeadingWord

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
move fct

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

OlivierBBB committed:
fix: return data in HALTING instructions + ras

letypequividelespoubelles committed:
je sais plus

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

OlivierBBB committed:
ras: renaming

# Conflicts:
#	arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/mmu/MmuCall.java

letypequividelespoubelles committed:
tracing of mmuCall.referenceSize

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
small modif prc subsection

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles reviewed (approved):
LGTM

OlivierBBB committed:
fix: NPE for precompiles

OlivierBBB committed:
Merge branch 'arith-dev' into 1547-unifying-ranges

letypequividelespoubelles committed:
Merge branch 'arith-dev' into 1547-unifying-ranges

OlivierBBB committed:
fix: update current return data whenever one updateCurrentReturnData()

OlivierBBB committed:
ras: slight streamlining

OlivierBBB committed:
spotless

OlivierBBB committed:
fix: rationalizing the creation of ContextFragments

There are now 3 main methods for creating ContextFragments:

- readContextFragment(Hub, CallFrame)
- updateReturnData(Hub, CallFrame, MemoryRange)
- initializeExecutionContext(Hub)

OlivierBBB committed:
ras

OlivierBBB committed:
fix: more snapshots

not sure if necessary ... ?

OlivierBBB committed:
fix: stupid mistake

likely introduced recently.

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/Hub.java:
```diff
@@ -1033,7 +1033,7 @@ void traceOpcode(MessageFrame frame) {
         }
 
         final boolean returnFromDeployment =
-            (this.opCode() == RETURN || this.currentFrame().isDeployment());
```
No idea what happened here 😆
